blaze/expr/arithmetic.py

```diff
@@ -50,18 +50,38 @@ def name(o):
         return None
 
 
-class BinOp(ElemWise):
-    __slots__ = '_hash', 'lhs', 'rhs'
+class Op(ElemWise):
+    __slots__ = '_hash', 'operands'
+    __inputs__ = 'operands'
+
+    def __init__(self, operands):
+        self._operands=operands
+
+    @property
+    def _inputs(self):
+        result = []
+        for _op in self._operands:
+            if isinstance(_op, Expr):
+               result.append(_op)
+        return tuple(result)
+
+class BinOp(Op):
+    __slots__ = '_hash'
     __inputs__ = 'lhs', 'rhs'
 
     def __init__(self, lhs, rhs):
-        self.lhs = lhs
-        self.rhs = rhs
+        super(BinOp, self).__init__(self, [lhs, rhs])
```

Comment:
this should be `super(BinOp, self).__init__([lhs, rhs])`. my apologies if my original suggestion wasn't this